### PR TITLE
tidb_query: fix CAST as string inconsistency between TiKV and TiDB

### PR DESCRIPTION
**Pull request by H-ZeX**
Signed-off-by: H-ZeX [hzx20112012@gmail.com](mailto:hzx20112012@gmail.com)

 * [x]  merge #5423 first. After merge #5423, the most of this PR is unit test, so it is easy to review, althouth the changed line number may seems very large.
 
### What have you changed?
compare tidb and tikv's impl of `cast` and fix what they are inconsistent.
There is more info of their difference in [this doc](https://hackmd.io/-3Nk55VVRtG8vuYCV3Lxiw#TiDB%E7%9A%84%E9%97%AE%E9%A2%98)

### What is the type of the changes?
* Improvement (a change which is an improvement to an existing feature)
* Bugfix (a change which fixes an issue)
* Engineering (engineering change which doesn't change any feature or fix any issue)

### How is the PR tested?
unit test

### Does this PR affect documentation (docs) or should it be mentioned in the release notes?
No.

### Does this PR affect `tidb-ansible`?
No.

### CODE REVIEW

File: components/tidb_query/src/codec/convert.rs [hunk 1]
```diff
@@ -711,8 +710,7 @@ pub fn pad_zero_for_binary_type(s: &mut Vec<u8>, ft: &FieldType) {
         return;
     }
     let flen = flen as usize;
-    if ft.as_accessor().tp() == FieldTypeTp::String && ft.is_binary_string_like() && s.len() < flen
-    {
+    if ft.tp() == FieldTypeTp::String && ft.is_binary_string_like() && s.len() < flen {
```
Comment (breeswish) [inline components/tidb_query/src/codec/convert.rs]: Use `.as_accessor()`.

File: components/tidb_query/src/rpn_expr/impl_cast.rs [hunk 2]
```diff
@@ -323,6 +333,159 @@ fn cast_json_as_uint(ctx: &mut EvalContext, val: &Option<Json>) -> Result<Option
...
+#[inline]
+fn cast_double_real_as_string(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra,
+    val: &Option<Real>,
+) -> Result<Option<Bytes>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            let val = val.into_inner();
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_cast.rs]: can convert being used?

Comment (H-ZeX) [reply components/tidb_query/src/rpn_expr/impl_cast.rs]:
```rust
    let res = produce_str_with_specified_tp(
        ctx,
        Cow::Borrowed(val.as_slice()),
        extra.ret_field_type,
        false,
    )?;
    let mut res = match res {
        Cow::Borrowed(_) => val,
        Cow::Owned(x) => x.to_vec(),
    };
    pad_zero_for_binary_type(&mut res, extra.ret_field_type);
    Ok(Some(res))
```

`cast_string_helper` need `extra.ret_field_type`, so it can not do by convert trait now.

File: components/tidb_query/src/rpn_expr/impl_cast.rs [hunk 3]
```diff
@@ -323,6 +333,159 @@ fn cast_json_as_uint(ctx: &mut EvalContext, val: &Option<Json>) -> Result<Option
     }
 }
 
+// cast any as string, some cast functions reuse `cast_any_as_any`
+//
+// cast_json_as_string -> by cast_any_as_any<Json, String>
+
+#[rpn_fn(capture = [ctx, extra])]
+#[inline]
+fn cast_int_as_string(
```
Comment (lonng) [inline components/tidb_query/src/rpn_expr/impl_cast.rs]: Can we define a macro to simplify this logic? (`cast_xxx_as_string` contains many duplicated logic)

File: components/tidb_query/src/codec/convert.rs [hunk 4]
```diff
@@ -711,8 +710,8 @@ pub fn pad_zero_for_binary_type(s: &mut Vec<u8>, ft: &FieldType) {
         return;
     }
     let flen = flen as usize;
-    if ft.as_accessor().tp() == FieldTypeTp::String && ft.is_binary_string_like() && s.len() < flen
-    {
+    let tp = ft.as_accessor().tp();
```
Comment (lonng) [inline components/tidb_query/src/codec/convert.rs]: Prefer to revert this change.

File: components/tidb_query/src/rpn_expr/impl_cast.rs [hunk 5]
```diff
@@ -2137,4 +2230,593 @@ mod tests {
...
+            ),
+            (
+                Json::Double(u64::MAX as f64),
+                format!("{:e}", u64::MAX as f64),
+            ),
+            (Json::Double(10.5), "10.5".to_string()),
+            (Json::Double(10.4), "10.4".to_string()),
+            (Json::Double(-10.4), "-10.4".to_string()),
+            (Json::Double(-10.5), "-10.5".to_string()),
+            (Json::String(String::from("10.0")), "\"10.0\"".to_string()),
```
Comment (lonng) [inline components/tidb_query/src/rpn_expr/impl_cast.rs]:
```suggestion
            (Json::String(String::from("10.0")), r#""10.0"".to_string()),
```

File: components/tidb_query/src/codec/convert.rs [hunk 6]
```diff
@@ -710,8 +710,8 @@ pub fn pad_zero_for_binary_type(s: &mut Vec<u8>, ft: &FieldType) {
         return;
     }
     let flen = flen as usize;
-    let tp = ft.as_accessor().tp();
-    if tp == FieldTypeTp::String && ft.is_binary_string_like() && s.len() < flen {
+    if ft.as_accessor().tp() == FieldTypeTp::String && ft.is_binary_string_like() && s.len() < flen
```
Comment (breeswish) [inline components/tidb_query/src/codec/convert.rs]: Maybe you want `ft.collation() == Collation::Binary`?

is_binary_string_like stands for `is_binary` && `is_string_like`. But you only want `is_binary` && `is_string`.

Comment (H-ZeX) [reply components/tidb_query/src/codec/convert.rs]: Yes, because we judge `ft.as_accessor().tp() == FieldTypeTp::String` first, then the `IsString(ft.Tp)` will always be true.
```go
func IsBinaryStr(ft *FieldType) bool {
	return ft.Collate == charset.CollationBin && IsString(ft.Tp)
}
```

I will fix this.
